Add HTML email layout

hursey013 opened the pull request:
**Why**:
- Add branding to application generated emails.

**How**:
- Template uses Zurb Foundation for Emails table based layout for email client compatibility, then inlines CSS styles using premailer-rails gem.  This allows small adjustments to be made to the layout without having to dig through a large HTML file with inlined CSS.

hursey013 commented:
@monfresh this is the PR I talked to you about - there is a doctype declaration in the top of the email layout which causes tests with `click_first_link_in_email` to fail.

sbc100 commented on Gemfile:
```diff
@@ -31,6 +31,7 @@ gem 'two_factor_authentication', github: 'Houdini/two_factor_authentication'
 gem 'uglifier', '>= 1.3.0'
 gem 'whenever', require: false
 gem 'xmlenc', '~> 0.6.4'
+gem 'premailer-rails'
```
Can you alphabetize this?

sbc100 commented on app/assets/stylesheets/email/_settings.scss:
```diff
@@ -0,0 +1,148 @@
+$thumbnail-transition: box-shadow 200ms ease-out;
+$thumbnail-radius: $global-radius;
+
+
```
Remove trailing lines

brendansudol replied:
let's name this file `variables` to be consistent with web stylez

sbc100 commented on app/mailers/custom_devise_mailer.rb:
```diff
@@ -1,4 +1,5 @@
 class CustomDeviseMailer < Devise::Mailer
+  layout 'layouts/user_mailer'
```
Add newline

brendansudol commented on app/views/layouts/user_mailer.html.slim:
```diff
@@ -0,0 +1,92 @@
+            table.spacer.float-center
+              tbody
+                tr
+                  td height="10px" style="font-size:10px;line-height:10px;"   
```
what's up with these inline styles?

hursey013 replied:
These are for creating vertical spacers in the email - since the size is arbitrary there really isn't a way to create a class for them unless we make a few utility classes with predefined sizes (.vs10, .vs25, etc) the problem is that I believe the td height tag needs to be explicitly set as well.  Reference: http://foundation.zurb.com/emails/docs/spacer.html

brendansudol replied:
cool, thanks; i would be in favor of adding these utility classes in a future PR

brendansudol commented on vendor/assets/stylesheets/foundation/foundation.scss:
```diff
@@ -0,0 +1,27 @@
+  'components/outlook-first',
+  'components/media-query';
+
+
```
Remove trailing lines

sbc100 commented on app/views/layouts/user_mailer.html.slim:
```diff
@@ -0,0 +1,92 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
```
Apparently you can just use `doctype strict`.

http://www.rubydoc.info/gems/slim/frames#__DOCTYPE__declaration

Do you know what the default is or why its not what we want?

brendansudol replied:
it looks like to minimize email rendering quirkiness, this doctype is actually preferred: https://www.campaignmonitor.com/blog/email-marketing/2010/11/correct-doctype-to-use-in-html-email/

sbc100 replied:
OK, but we can still use the shorthand version can't we?

monfresh commented on app/views/email_second_factor_mailer/your_code_is.html.slim:
```diff
@@ -1,13 +1,7 @@
+  'If you are not attempting to log into #{APP_NAME}, please contact the Customer Contact Center by filling out the web form at
+  = link_to nil, Figaro.env.support_url
+  |  for assistance.
+p Please be aware that this update might require your immediate attention.
```
This entire file can be removed. We are not sending OTPs via email.

monfresh commented on app/views/layouts/user_mailer.html.slim:
```diff
@@ -0,0 +1,92 @@
+                                        p.text-center
+                                          a href="#" About #{APP_NAME} 
+                                          ' | 
+                                          a href="#" Privacy policy
```
Are these links placeholders for now or do we have About and Privacy policy pages we can link to?

hursey013 replied:
Placeholders right now, we do not have these pages created quite yet.

monfresh commented on app/views/user_mailer/password_expiry.html.slim:
```diff
@@ -1,16 +1,8 @@
-    p
-      strong PLEASE DO NOT REPLY TO THIS MESSAGE
+p
+  = link_to @link_url, @link_url, target: '_blank'
```
This entire file should be removed. We are no longer expiring passwords.

monfresh commented on vendor/assets/stylesheets/foundation/components/_type.scss:
```diff
@@ -0,0 +1,316 @@
+  }
+}
+
+
```
Do we need all of this CSS? Can we eliminate anything that's not being used?

monfresh commented:
Is there a particular reason we are adding another CSS framework? Does `foundation` provide something we can't do with basscss or custom CSS?

brendansudol commented:
I'm a bit torn on this too, @monfresh @hursey013, re: adding another css framework for email, but i think i'm ok with it for now. email styling is much different (aka worse) than web styling; certain things that work without question on the web (in any browser) fail dramatically for email clients. by using foundation for email (note: not the entire foundation css suite), we're able to mitigate a lot of these email styling issues. that said, i don't love there is some duplication happening here (regarding color, font size declarations, utility classes, etc.) and i think we should try to unify / minimize those soon (in a future PR).

monfresh commented on spec/views/devise/mailer/confirmation_instructions.html.slim_spec.rb:
```diff
@@ -11,10 +11,7 @@
-    render
-
-    expect(rendered).to have_link(Figaro.env.support_url, href: Figaro.env.support_url)
+    to include Figaro.env.support_url
```
This seems to be a copy and paste error. Looks like the `expect(rendered)` got dropped off. Also, was there a reason you removed the first part of this test?

hursey013 replied:
@monfresh This test is failing because the link to the support url has been moved into the user_mailer layout.  This test seems to only look at the rendered content of the actual confirmation_instructions.html.slim file which no longer includes the support url.  I used last_email in order to see the entire email, but that may not be the correct way to do it.

monfresh commented on spec/views/devise/mailer/confirmation_instructions.html.slim_spec.rb:
```diff
@@ -29,10 +26,8 @@
-
-    expect(rendered).to have_content 'PLEASE DO NOT REPLY TO THIS MESSAGE'
+    expect(last_email.parts.first.body.raw_source).
+      to have_content 'Please do not reply to this message.'
```
If you run this test on its own, it will fail because there is no `last_email` variable in a view spec. No emails are being sent in view specs. I think the only change that needs to be made here is the expected text.

hursey013 replied:
@monfresh This test is failing because the text "please do not reply to this message" has been moved into the user_mailer layout.  This test seems to only look at the rendered content of the actual confirmation_instructions.html.slim file which no longer includes the text.  I used last_email in order to see the entire email, but that may not be the correct way to do it.

monfresh commented on app/views/layouts/user_mailer.html.slim:
```diff
@@ -0,0 +1,92 @@
+                            table
+                              tr
+                                th
+                                  = image_tag("logo.png", class: "small-float-center")
```
This will not render the image because it does not produce a full URL. To fix this, we need to use `asset_url` with a `host` parameter, such as:
```
= image_tag(asset_url("logo.png", host: Figaro.env.domain_name), class: "small-float-center")
```
The problem is that the `host` option needs a protocol as well, and our `domain_name` entries in `application.yml` do not include the protocol. This will need to be fixed in a separate PR, but in the meantime, you can create a new entry in `application.yml.example` called `mailer_domain_name` and set it to `http://localhost:3000` for development. Then you can use it for the image by calling `Figaro.env.mailer_domain_name`

monfresh replied:
Also, we already have a logo.svg. Can that be reused here, or do we need a separate PNG?

hursey013 replied:
SVG support in various email clients is still spotty, so using a PNG is likely the most reliable method.  Reference: https://css-tricks.com/a-guide-on-svg-support-in-email/

monfresh commented:
Code Climate does not currently have an engine for linting Slim files, so could you please run `slim-lint` for all the files you modified and fix the offenses? For example, to see offenses in the `user_mailer` layout file you added, you would run:
```
slim-lint app/views/layouts/user_mailer.html.slim
```
Repeat for each file you added or modified.

Also, note that we use Overcommit to display those errors for you so you don't have to run slim-lint file by file. To make sure Overcommit is enabled, run `bin/setup`, and when you create a commit that includes Slim files with offenses, you should see some output from Overcommit reporting the Slim errors.

hursey013 commented:
@monfresh I have addressed all of the issues with the exception of the click_first_link_in_email test catching the doctype url instead of the first actual link in the emails.  There is also a test to ensure that the support url and "please do not reply" text are in the email body - but since those are now permanently in the user_mailer layout the test needs to be modified or possibly removed.

sbc100 commented on app/views/layouts/user_mailer.html.slim:
```diff
@@ -0,0 +1,95 @@
+    /! prevent Gmail on iOS font size manipulation
+    div style="display:none; white-space:nowrap; font:15px courier; line-height:0;"
+      |
+        &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0; &#xA0;
```
What are all these things?

File should end in a newline.

Also, I think you can put the `|` and the content on the same line.

hursey013 replied:
@sbc100 These spaces are necessary in order to prevent gmail on iOS from automatically enlarging all of the text and messing up the layout.  I bumped them up onto the same line as the `|` but for some reason the last line is automatically stripped.  My local copy has a line return.  Looks like the main application layout is the same way: https://github.com/18F/identity-idp/blob/master/app/views/layouts/application.html.slim

sbc100 commented on app/views/layouts/user_mailer.html.slim:
```diff
@@ -0,0 +1,95 @@
+                                    tr
+                                      th
+                                        p
+                                          'For more help, please contact the #{APP_NAME} Customer Contact Center via web form at
```
Use `|` here for consistency?

sbc100 commented on app/views/layouts/user_mailer.html.slim:
```diff
@@ -0,0 +1,95 @@
+                                      th
+                                        p.text-center
+                                          a href="#" About #{APP_NAME}
+                                          ' |
```
Does `| |` work here?

monfresh commented:
I will checkout this branch locally and work on getting the tests to pass

monfresh commented:
Hi @hursey013. To fix the password recovery specs, you can replace `click_first_link_in_email` with the `click_email_link_matching` helper, which takes a regex as a parameter. To fix the confirmation instructions spec, I removed those expectations and added them to a new spec for the mailer layout. 

I just pushed a commit that fixes the tests.

hursey013 commented:
awesome, thank you for the assistance @monfresh
